server/mediaserver.go

```diff
@@ -489,12 +310,84 @@ func endRTMPStreamHandler(s *LivepeerServer) func(url *url.URL, rtmpStrm stream.
 		if monitor.Enabled {
 			monitor.LogStreamEndedEvent(cxn.nonce)
 		}
+		cxn.eof <- struct{}{}
 		delete(s.rtmpConnections, mid)
 
 		return nil
 	}
 }
 
+func (s *LivepeerServer) startSession(cxn *rtmpConnection) *BroadcastSession {
+
+	mid := rtmpManifestID(cxn.stream)
+	cpl := cxn.pl
+	var sess *BroadcastSession
+
+	// Connect to the orchestrator. If it fails, retry for as long
+	// as the RTMP stream is alive; maybe the orchestrator hasn't
+	// received the block containing the job yet
+	broadcastFunc := func() error {
+		var err error
+		sess, err = selectOrchestrator(s.LivepeerNode, cpl)
+		if err == ErrDiscovery {
+			return nil // discovery disabled, don't retry
+		} else if err != nil {
+			// Should be logged upstream
+		}
+		s.connectionLock.RLock()
+		defer s.connectionLock.RUnlock()
+		_, active := s.rtmpConnections[mid]
+		if active {
+			return err
+		}
+		return nil // stop if inactive
+	}
+	expb := backoff.NewExponentialBackOff()
+	expb.MaxInterval = BroadcastRetry
+	expb.MaxElapsedTime = 0
+	backoff.Retry(broadcastFunc, expb)
+	return sess
+}
+
+func (s *LivepeerServer) startSessionListener(cxn *rtmpConnection) {
+	mut := sync.Mutex{}
+	cv := sync.NewCond(&mut)
+	pending := false
+	runStartSession := func() {
+		sess := s.startSession(cxn)
+		mut.Lock()
+		defer mut.Unlock()
```

Comment:
For my own info - when we `defer` here, does the mutex Unlock at the end of runStartSession? Or at the end of startSessionListener?

Reply:
Yes, the mutex unlocks at the end of `runStartSession` ; defers are function scoped. Defer actually has a lot of gotchas which is unfortunate because it's a neat construct that minimizes some common errors such as missing a clean-up step (such as unlocking a mutex) in a error-return path.

See https://blog.learngoprogramming.com/gotchas-of-defer-in-go-1-8d070894cb01

Reply:
What a great article! Thank you Josh